sorts/tim_sort.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * Comparator function type definition.
+ *
+ * @typeparam T The type of elements being compared.
+ */
+type Comparator<T> = (a: T, b: T) => number
+
+// Minimum size of subarrays to be sorted using insertion sort before merging
+const MIN_MERGE = 32
+
+/**
+ * Performs insertion sort on a portion of an array.
+ *
+ * @typeparam T The type of elements in the array.
+ * @param arr The array to sort.
+ * @param startIndex The start index of the portion to sort.
+ * @param endIndex The end index of the portion to sort.
+ * @param compare The comparator function defining the order of elements.
+ */
+const insertionSort = <T>(
+  arr: T[],
+  startIndex: number,
+  endIndex: number,
+  compare: Comparator<T>
+): void => {
+  for (let i = startIndex + 1; i <= endIndex; i++) {
+    const currentElement = arr[i]
+    let j = i - 1
+
+    while (j >= startIndex && compare(arr[j], currentElement) > 0) {
+      arr[j + 1] = arr[j]
+      j--
+    }
+    arr[j + 1] = currentElement
+  }
+}
+
+/**
+ * Merges two sorted subarrays into one sorted array.
+ *
+ * @typeparam T The type of elements in the array.
+ * @param arr The array containing the subarrays to merge.
+ * @param leftIndex The left index of the first subarray.
+ * @param middleIndex The middle index separating the two subarrays.
+ * @param rightIndex The right index of the second subarray.
+ * @param compare The comparator function defining the order of elements.
+ */
+const merge = <T>(
+  arr: T[],
+  leftIndex: number,
+  middleIndex: number,
+  rightIndex: number,
+  compare: Comparator<T>
+): void => {
+  const leftArrayLength = middleIndex - leftIndex + 1
+  const rightArrayLength = rightIndex - middleIndex
+
+  const leftArray: T[] = arr.slice(leftIndex, middleIndex + 1)
+  const rightArray: T[] = arr.slice(middleIndex + 1, rightIndex + 1)
+
+  let leftPointer = 0
+  let rightPointer = 0
+  let mergedIndex = leftIndex
+
+  while (leftPointer < leftArrayLength && rightPointer < rightArrayLength) {
+    if (compare(leftArray[leftPointer], rightArray[rightPointer]) <= 0) {
+      arr[mergedIndex++] = leftArray[leftPointer++]
+    } else {
+      arr[mergedIndex++] = rightArray[rightPointer++]
+    }
+  }
+
+  while (leftPointer < leftArrayLength) {
+    arr[mergedIndex++] = leftArray[leftPointer++]
+  }
+
+  while (rightPointer < rightArrayLength) {
+    arr[mergedIndex++] = rightArray[rightPointer++]
+  }
+}
+
+/**
+ * Sorts an array using the Tim sort algorithm.
+ *
+ * @typeparam T The type of elements in the array.
+ * @param arr The array to sort.
+ * @param compare The comparator function defining the order of elements.
+ */
+export const timSort = <T>(arr: T[], compare: Comparator<T>): void => {
+  const n = arr.length
+
+  for (let i = 0; i < n; i += MIN_MERGE) {
+    insertionSort(arr, i, Math.min(i + MIN_MERGE - 1, n - 1), compare)
+  }
+
+  for (let size = MIN_MERGE; size < n; size *= 2) {
```

Comment:
I can see that you tried to stick to the "core idea" of Timsort - which is to leverage sorted runs, and to use insertion sort to produce small sorted runs to start with, since insertion sort may have better constant factors on short slices - but this does not look like a proper Timsort to me.

At this point, you have sorted runs of constant length `MIN_MERGE`, and it looks like what you're doing now is just an iterative mergesort.

This gives you Θ(n log n) time complexity, irrespective of the input. It defeats a defining characteristic of Timsort: To have better runtime if the data allows it.

Timsort achieves this by managing a stack of sorted runs with special size invariants in order to do balanced merges ([source](https://en.wikipedia.org/wiki/Timsort#Merge_criteria). I see no such stack here.

For example, in the best case where the data is already sorted, you should have a single sorted run and no merging should happen; you should achieve O(n) time complexity. I don't see how this is implemented here.

Furthermore, handling of descending runs seems to be lacking.